File: JGSourceBase/JGSourceBase/JGSSafe/JGSProxyDetector.m
//
//  JGSProxyDetector.m
//  JGSourceBase
//
//  Created by Mei JiGao on 2026/6/25.
//  Copyright © 2026 ByMountains. All rights reserved.
//

#import "JGSProxyDetector.h"
#import <CFNetwork/CFNetwork.h>
#import "JGSLogger+Private.h"

@implementation JGSProxyDetector

/// 系统层VPN、网络代理开启状态判断，无法针对特定域名判断
+ (JGSProxyOptions)proxyEnabledTypes {
    // 系统代理设置
    NSDictionary *settings = (__bridge_transfer NSDictionary *)CFNetworkCopySystemProxySettings();
    if (!settings) {
        return kNilOptions;
    }
    JGSDebugLog(@"settings: %@", settings);

    // 网络代理设置
    NSDictionary *proxyTypes = @{
        @"HTTPEnable": @(JGSProxyOptionsHTTP),
        @"HTTPSEnable": @(JGSProxyOptionsHTTPS),
        @"SOCKSEnable": @(JGSProxyOptionsSOCKS),
        @"ProxyAutoConfigEnable": @(JGSProxyOptionsAuto),
    };
    JGSProxyOptions options = kNilOptions;
    for (NSString *key in proxyTypes) {
        id enable = settings[key];
        if ([enable isKindOfClass:[NSNumber class]] && [enable isEqual:@(1)]) {
            options |= [proxyTypes[key] integerValue];
        }
    }

    // VPN设置
    // 当 VPN 连接时，系统会创建一个虚拟网络接口（如 tun0、utun1），这个接口信息会出现在 __SCOPED__ 字典中
    NSDictionary *scopeSettings = settings[@"__SCOPED__"];
    if ([scopeSettings isKindOfClass:[NSDictionary class]]) {
        // VPN 接口关键词：VPN 连接会创建虚拟网络接口，其名称通常包含以下关键词
        // tap / tun：最常见的 VPN 隧道接口
        // ppp：点对点协议，用于某些 VPN 类型
        // ipsec：IPsec VPN 协议
        // utun：用户态 tun 接口，在较新的 iOS 系统中被广泛使用
        NSArray<NSString *> *vpnKeys = @[@"tap", @"tun", @"ppp", @"ipsec", @"utun"];
        BOOL vpnDetected = NO;
        for (NSString *vpnKey in vpnKeys) {
            for (NSString *scopeKey in scopeSettings.allKeys) {
                if ([scopeKey hasPrefix:vpnKey]) {
                    vpnDetected = YES;
                    break;
                }
            }
            if (vpnDetected) break;
        }
        if (vpnDetected) {
            options |= JGSProxyOptionsVPN;
        }
    }

    JGSDebugLog(@"enableTypes: %zd", options);
    return options;
}

/// 系统层VPN开启状态判断（不针对特定URL）、特定URL的网络代理配置
+ (JGSProxyOptions)proxyEnabledTypesFor:(NSString *)url {
    NSURL *URL = [NSURL URLWithString:url];
    if (!URL) {
        return [self proxyEnabledTypes];
    }
    return [self proxyEnabledTypesForURL:URL];
}

/// 系统层VPN开启状态判断（不针对特定URL）、特定URL的网络代理配置
+ (JGSProxyOptions)proxyEnabledTypesForURL:(NSURL *)url {
    // 系统代理设置
    CFDictionaryRef cfSettings = CFNetworkCopySystemProxySettings();
    if (!cfSettings) {
        return kNilOptions;
    }

    NSArray *proxies = (__bridge_transfer NSArray *)CFNetworkCopyProxiesForURL((__bridge CFURLRef)url, cfSettings);
    NSDictionary *settings = (__bridge_transfer NSDictionary *)cfSettings;
    JGSDebugLog(@"settings: %@", settings);
    JGSDebugLog(@"proxies for %@ : %@", url, proxies);
    
    // 提取代理类型
    NSMutableSet *urlProxyTypes = [NSMutableSet set];
    for (NSDictionary *proxy in proxies) {
        if ([proxy isKindOfClass:[NSDictionary class]]) {
            NSString *type = proxy[(__bridge NSString *)kCFProxyTypeKey];
            if (type) {
                [urlProxyTypes addObject:type];
            }
        }
    }

    // 网络代理设置
    NSDictionary *proxyTypes = @{
        (__bridge NSString *)kCFProxyTypeHTTP: @(JGSProxyOptionsHTTP),
        (__bridge NSString *)kCFProxyTypeHTTPS: @(JGSProxyOptionsHTTPS),
        (__bridge NSString *)kCFProxyTypeSOCKS: @(JGSProxyOptionsSOCKS),
        (__bridge NSString *)kCFProxyTypeAutoConfigurationURL: @(JGSProxyOptionsAuto),
    };
    JGSProxyOptions options = kNilOptions;
    for (NSString *key in proxyTypes) {
        if ([urlProxyTypes containsObject:key]) {
            options |= [proxyTypes[key] integerValue];
        }
    }

    // VPN设置
    // 当 VPN 连接时，系统会创建一个虚拟网络接口（如 tun0、utun1），这个接口信息会出现在 __SCOPED__ 字典中
    NSDictionary *scopeSettings = settings[@"__SCOPED__"];
    if ([scopeSettings isKindOfClass:[NSDictionary class]]) {
        // VPN 接口关键词
        NSArray<NSString *> *vpnKeys = @[@"tap", @"tun", @"ppp", @"ipsec", @"utun"];
        BOOL vpnDetected = NO;
        for (NSString *vpnKey in vpnKeys) {
            for (NSString *scopeKey in scopeSettings.allKeys) {
                if ([scopeKey hasPrefix:vpnKey]) {
                    vpnDetected = YES;
                    break;
                }
            }
            if (vpnDetected) break;
        }
        if (vpnDetected) {
            options |= JGSProxyOptionsVPN;
        }
    }

    JGSDebugLog(@"enableTypes for %@ : %zd", url, options);
    return options;
}

@end

File: JGSourceBase/JGSourceBase/JGSSafe/JGSProxyDetector.swift
//
//  JGSProxyDetector.swift
//  JGSourceBase
//
//  Created by Mei JiGao on 2026/6/23.
//  Copyright © 2026 ByMountains. All rights reserved.
//

import CFNetwork

public
struct JGSProxyOptions: OptionSet, @unchecked Sendable {
    public let rawValue: Int
    public init(rawValue: Int) {
        self.rawValue = rawValue
    }
    
    /// 网络配置：手动http代理配置
    public static var http: JGSProxyOptions { JGSProxyOptions(rawValue: 1 << 0) }
    /// 网络配置：手动https代理配置
    public static var https: JGSProxyOptions { JGSProxyOptions(rawValue: 1 << 1) }
    /// 网络配置：手动socks代理配置
    public static var socks: JGSProxyOptions { JGSProxyOptions(rawValue: 1 << 2) }
    /// 网络配置：自动代理配置/
    public static var auto: JGSProxyOptions { JGSProxyOptions(rawValue: 1 << 3) }
    /// VPN网络
    public static var vpn: JGSProxyOptions { JGSProxyOptions(rawValue: 1 << 4) }
}

public final
class JGSProxyDetector: JGSSwiftClass<NSObject> {
    
    /// 系统层VPN、网络代理开启状态判断，无法针对特定域名判断
    /// - Returns: 代理、VPN开启状态
    public static
    func proxyEnabledTypes() -> JGSProxyOptions {
        // 系统代理设置
        guard let settings = CFNetworkCopySystemProxySettings()?.takeRetainedValue() as? [String: Any] else {
            return []
        }
        JGSDebugLog("settings:", settings)
        
        // 网络代理设置
        let proxyTypes: [String: JGSProxyOptions] = [
            "HTTPEnable" : .http,
            "HTTPSEnable": .https,
            "SOCKSEnable": .socks,
            "ProxyAutoConfigEnable": .auto,
        ]
        var options: [JGSProxyOptions] = proxyTypes.compactMap({ (key: String, value: JGSProxyOptions) in
            // 测试验证发现存在以下情况：
            // 1. iPhone真机无代理设置，则不返回相关enable字段
            // 2. 模拟器无代理设置，返回相关enable字段可能为0
            // 因此，需要判断enable值
            if let enable = settings[key] as? Bool, enable {
                return value
            } else if let enable = settings[key] as? NSNumber, enable.isEqual(to: NSNumber(value: 0)) {
                return value
            }
            return nil
        })
        
        // VPN设置
        // 当 VPN 连接时，系统会创建一个虚拟网络接口（如 tun0、utun1），这个接口信息会出现在 __SCOPED__ 字典中
        if let scopeKeys = (settings["__SCOPED__"] as? [String: Any])?.keys {
            // VPN 接口关键词：VPN 连接会创建虚拟网络接口，其名称通常包含以下关键词
            // tap / tun：最常见的 VPN 隧道接口
            // ppp：点对点协议，用于某些 VPN 类型
            // ipsec：IPsec VPN 协议
            // utun：用户态 tun 接口，在较新的 iOS 系统中被广泛使用
            let vpnTypes = ["tap", "tun", "ppp", "ipsec", "utun"].filter { vpnKey in
                for scopeKey in scopeKeys {
                    // 接口名称通常带有数字后缀（如 tun0、utun1）
                    // 使用 starts 进行模糊匹配能更准确地识别
                    if scopeKey.starts(with: vpnKey) {
                        return true
                    }
                }
                return false
            }
            if vpnTypes.count > 0 {
                options.append(.vpn)
            }
        }
        
        let enableTypes = JGSProxyOptions(options)
        JGSDebugLog(format: "enableTypes: %@", enableTypes)
        
        return enableTypes
    }
    
    /// 系统层VPN开启状态判断（VPN状态无法针对特定URL）、特定URL的网络代理配置
    /// iPhone无法配置例外，可通过模拟器运行，Mac 网络代理配置 http、https、socks、例外进行验证
    /// - Parameter url: 指定url
    /// - Returns: VPN、代理开启状态
    public static
    func proxyEnabledTypes(for url: String) -> JGSProxyOptions {
        guard let URL = URL(string: url) else {
            return proxyEnabledTypes()
        }
        return proxyEnabledTypes(for: URL)
    }
    
    /// 系统层VPN开启状态判断（VPN状态无法针对特定URL）、特定URL的网络代理配置
    /// iPhone无法配置例外，可通过模拟器运行，Mac 网络代理配置 http、https、socks、例外进行验证
    /// - Parameter url 指定URL
    /// - Returns: VPN、代理开启状态
    public static
    func proxyEnabledTypes(for url: URL) -> JGSProxyOptions {
        // 系统代理设置
        guard let settings = CFNetworkCopySystemProxySettings()?.takeRetainedValue(),
              let proxies = CFNetworkCopyProxiesForURL(url as CFURL, settings).takeRetainedValue() as? [[CFString: Any]]
        else {
            return []
        }
        JGSDebugLog("settings:", settings)
        JGSDebugLog("proxies for \(url) :", proxies)
        
        let urlProxies = proxies.compactMap({ dict in
            return dict[kCFProxyTypeKey]
        }) as? [CFString] ?? []
        
        // 网络代理设置
        let proxyTypes: [CFString: JGSProxyOptions] = [
            kCFProxyTypeHTTP : .http,
            kCFProxyTypeHTTPS: .https,
            kCFProxyTypeSOCKS: .socks,
            kCFProxyTypeAutoConfigurationURL: .auto,
        ]
        var options: [JGSProxyOptions] = proxyTypes.compactMap({ (key: CFString, value: JGSProxyOptions) in
            if urlProxies.contains(key) {
                return value
            }
            return nil
        })
        
        // VPN设置
        // 当 VPN 连接时，系统会创建一个虚拟网络接口（如 tun0、utun1），这个接口信息会出现在 __SCOPED__ 字典中
        if let scopeKeys = ((settings as? [String: Any])?["__SCOPED__"] as? [String: Any])?.keys {
            // VPN 接口关键词：VPN 连接会创建虚拟网络接口，其名称通常包含以下关键词
            // tap / tun：最常见的 VPN 隧道接口
            // ppp：点对点协议，用于某些 VPN 类型
            // ipsec：IPsec VPN 协议
            // utun：用户态 tun 接口，在较新的 iOS 系统中被广泛使用
            let vpnTypes = ["tap", "tun", "ppp", "ipsec", "utun"].filter { vpnKey in
                for scopeKey in scopeKeys {
                    // 接口名称通常带有数字后缀（如 tun0、utun1）
                    // 使用 starts 进行模糊匹配能更准确地识别
                    if scopeKey.starts(with: vpnKey) {
                        return true
                    }
                }
                return false
            }
            if vpnTypes.count > 0 {
                options.append(.vpn)
            }
        }
        
        let enableTypes = JGSProxyOptions(options)
        JGSDebugLog("enableTypes for \(url):", enableTypes.rawValue)
        
        return enableTypes
    }
}
